### DISCUSSION TITLE
How to invalidate a list with arbitrary filters

### DISCUSSION THREAD

**Discussion by nandorojo** [urql-graphql/urql#2553] [Q&A] [answered]
This is a bit of a duplicate of https://github.com/FormidableLabs/urql/discussions/2214. I'm hoping to get a sense of how `cache.invalidate` works for lists.

I have a mutation called `createBooking`. Whenever it succeeds, I want to invalidate `myBookings`. `myBookings` is used in many screens with unknown filters (such as `startDateMax`, `limit`, `offset`, etc.)

Is this sufficient to invalidate every instance of `myBookings`, or do I need to `inspectFields`?

```ts
const Mutation = {
    createBooking(parent, args, cache, info) { 
       cache.invalidate('Query', 'myBookings')
    },
}
```

I ask because this is my current setup for graphcache, but when I create a booking, the list of bookings doesn't refresh. `myBookings` also has `simplePagination` as a resolver, in case that's useful info.

**Comment by JoviDeCroock** [comment, marked as answer]
This won't work, you will need to leverage `cache.inspectFields` when we assemble the key for the above call we will purge `Query.myBookings` which to your description never happened. I do think there might be value in getting an invalidate field disregard args type of invalidate but that isn't currently there.

```js
const fields = cache
  .inspectFields('Query')
  .filter(field => field.fieldName === 'myBookings')
  .forEach(field => {
    cache.invalidate('Query', 'myBookings', field.arguments)
  });
```

Slightly annoying to type but yes 😅

**Reply by nandorojo** [reply]
ahh, i see, good to know. the `inspectFields` example on the docs only showed manually updating the cache, but not invalidating it. thanks!!

**Reply by nandorojo** [reply]
For anyone who wants to loop fewer times / have good TS support, this is what I used:

```ts
function createBooking(parent, args, cache, info) {
  const fieldName = makeQueryKey('myBookings')

  for (const field of cache.inspectFields('Query')) {
    if (field.fieldName == fieldName) {
      cache.invalidate('Query', field.fieldName, field.arguments)
    }
  }
}
```

`makeQueryKey` imports `Query`, which is codegen'd by GraphQL Codegen's URQL plugin. It's just a typesafe way to generate a field name.

```ts
const makeQueryKey = <Key extends keyof Query>(key: Key): Key => key
```

**Reply by nandorojo** [reply]
It would be cool for `cache.invalidate` to have better TS autocomplete from the codegen :)
